Implemented order available payment methods and order update payment

dlobato opened the pull request:
Hi,

We have implemented two new actions to handle payment state on placed orders:

1. Get available payment methods: Same as /{channelCode}/checkout/{token}/payment but on order endpoint.
2. Update payment method on unpaid placed orders to implement the same behavior as in ShopBundle to allow a customer to change the payment method while order is unpaid.

1 is just reusing the same action from checkout, not sure if I should reimplement this as a different action.
2 adds validation as I've seen on other controllers, is this the preferred approach?

Tests are failing, there's something weird I couldn't figure out on loading the fixtures. I bet I'm doing something wrong. Am I missing something obvious there?

I've also added a new field to placedorder view: paymentState.

David.

mamazu commented on src/Validator/Constraints/OrderExists.php:
```diff
@@ -0,0 +1,44 @@
+        if (isset($options['state'])) {
+            $options['state'] = array_map(
+                static function ($i) {
+                    return defined($i) ? constant($i) : $i;
```
There is just a limited amour of possibilities, you could do a switch statement here.

dlobato replied:
If you think the constant translation is overkill, I can remove it and the code inside the if statement would look like:
`$options['state'] = (array)$options['state']`

Not sure why would we need and switch here, do you mean to validate thet state is a valid order state?

mamazu replied:
What should this do? The job of the constraint is not to translate the state it is given.

dlobato replied:
fair point, I'll simplify this

mamazu commented on tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php:
```diff
@@ -0,0 +1,73 @@
+        }
+EOT;
+
+        $this->client->request('PUT', $this->getPaymentUrl($placedOrder->getTokenValue()) . '/0', [], [], self::CONTENT_TYPE_HEADER, $data);
```
Just use the value that is in the fixture. No need to reference the object.

dlobato replied:
Sure, makes sense. Anyway, tests fails because:
ORDERTOKENPLACED has state cart, which hits on the validation rule. But if I change it to new, then other tests (order related) fail because totals don't match anymore... Maybe I should create another test order.

But the second test (ORDERTOKENPAID) throws an exception while loading fixtures, and I can't figure out why, have you ever seen this? I'm doing something wrong, sure, but where?

mamazu replied:
Yeah creating a new order for that matter would be a good idea to not interfere with other tests.

About the fixture loading, I haven't seen this issue. But I can try to look into it.

dlobato replied:
There two new orders with required information.

dlobato replied:
https://github.com/Sylius/ShopApiPlugin/pull/422/files#diff-2170a1f75ea7146d4434637215da40b7

GSadee commented on src/Command/Order/UpdatePaymentMethod.php:
```diff
@@ -0,0 +1,39 @@
+    protected $orderToken;
+
+    /** @var mixed */
+    protected $paymentIdentifier;
```
`$paymentId`?

dlobato replied:
copy paste from here \Sylius\ShopApiPlugin\Command\Cart\ChoosePaymentMethod::$paymentIdentifier

GSadee commented on src/Command/Order/UpdatePaymentMethod.php:
```diff
@@ -0,0 +1,39 @@
+    protected $paymentIdentifier;
+
+    /** @var string */
+    protected $paymentMethod;
```
`$paymentMethodCode` to be more clear?

dlobato replied:
same as above: \Sylius\ShopApiPlugin\Command\Cart\ChoosePaymentMethod::$paymentMethod
but sure I can update those

GSadee replied:
I would be for changing that in the added classes in this PR (and open a new PR for the changes in existing classes)

GSadee commented on src/Handler/Order/UpdatePaymentMethodHandler.php:
```diff
@@ -0,0 +1,49 @@
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $choosePaymentMethod->orderToken()]);
+
+        Assert::notNull($order, 'Order has not been found.');
+        Assert::eq(OrderPaymentStates::STATE_AWAITING_PAYMENT, $order->getPaymentState(), 'Only awaiting payment orders can be updated.');
```
`eq` => `same`

GSadee commented on src/Resources/config/routing/order.yml:
```diff
@@ -9,3 +9,15 @@ sylius_shop_api_order_details:
+    path: /orders/{token}/payment/{paymentId}
+    methods: [PUT]
+    defaults:
+        _controller: sylius.shop_api_plugin.controller.order.update_payment_method_action
```
Missing blank line at the end of file

GSadee commented on src/Validator/Constraints/OrderExists.php:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    /** {@inheritdoc} */
+    public function getTargets()
```
You can add `string` return type and docblock is not necessary IMO

GSadee commented on tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php:
```diff
@@ -0,0 +1,67 @@
+        $this->logInUser('hater@queen.com', '123password');
+
+        $data =
+            <<<EOT
```
Wrong indentation

GSadee commented on tests/Controller/Order/OrderUpdatePaymentMethodApiTest.php:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @return string
+     */
+    private function getPaymentUrl($token): string
```
`private function getPaymentUrl(string $token): string` and docblock is not needed

dlobato commented:
@GSadee @mamazu I'm struggling here with the tests... Can't figure out what's wrong. If I run the tests alone on local they work fine, but when running them one after the other there's an unrelated error. Seems to me some sort of race condition while loading the fixtures for the second time. Is something obvious that I'm not seeing here?

mamazu commented on src/Validator/Constraints/OrderExists.php:
```diff
@@ -13,28 +13,9 @@ final class OrderExists extends Constraint
     public $message = 'sylius.shop_api.order.not_exists';
 
-    /** @var array */
+    /** @var mixed */
```
Why did you change that? Isn't the value an array?

dlobato replied:
No, it can be just a string or an array of strings. The constraint mapping xml accepts this two ways of passing options:
```
<option name="state">new</option>
```
or
```
<option name="state">
   <value>new</value>
   <value>other</value>
</option>
```

mamazu replied:
Yeah, I already feared so then I would suggest: `/** @var string|array<string> */` so that we have at least some kind of idea what we are dealing with (and it helps the static analysis).

dlobato commented:
I fixed the tests by changing the item used on the order. Sadly, I couldn't figure out what was the problem but it is related to mug item having product attributes. Since the items on the order are irrelevant for the test purposes I've just used another one and updated the expected responses accordingly.

@mamazu @GSadee is there anything else to code review?

mamazu commented:
Ill have a look at it this evening.

mamazu commented on spec/Handler/Order/UpdatePaymentMethodHandlerSpec.php:
```diff
@@ -0,0 +1,119 @@
+        $this->beConstructedWith($orderRepository, $paymentMethodRepository);
+    }
+
+    function it_assigns_choosen_payment_method_to_specified_payment(
```
choosen => chosen

mamazu commented on src/Handler/Order/UpdatePaymentMethodHandler.php:
```diff
@@ -0,0 +1,49 @@
+    public function __invoke(UpdatePaymentMethod $choosePaymentMethod): void
+    {
+        /** @var OrderInterface $order */
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $choosePaymentMethod->orderToken()]);
```
This call will also return carts. You need to check the state as well.

dlobato replied:
Now I check order state is not cart. I've updated phpspec tests accordingly.

mamazu commented:
There are still some open conversations if you could resolve that, this would be great.

lchrusciel commented on doc/swagger.yml:
```diff
@@ -1009,6 +1009,53 @@ paths:
                     description: "Order with given tokenValue not found"
             security:
                 -   bearerAuth: []
+    /{channelCode}/orders/{token}/payment:
```
`payment` -> `payments` ?

dlobato replied:
I'd rather keep it like this to be homogeneous with /{channelCode}/checkout/{token}/payment

lchrusciel commented on spec/Command/Order/UpdatePaymentMethodSpec.php:
```diff
@@ -0,0 +1,45 @@
+        $this->shouldThrow(TypeError::class)->duringInstantiation();
+    }
+
+    function it_throws_an_exception_if_payment_method_code_is_not_a_string(): void
```
Since the introduction of strict type, we can remove TypeError specs

lchrusciel commented on spec/Handler/Order/UpdatePaymentMethodHandlerSpec.php:
```diff
@@ -0,0 +1,143 @@
+use Sylius\Component\Core\Repository\PaymentMethodRepositoryInterface;
+use Sylius\ShopApiPlugin\Command\Order\UpdatePaymentMethod;
+
+class UpdatePaymentMethodHandlerSpec extends ObjectBehavior
```
specs should be final

lchrusciel commented on tests/DataFixtures/ORM/order.yml:
```diff
@@ -49,6 +70,11 @@ Sylius\Component\Core\Model\Payment:
         order: "@placed_order"
         currencyCode: "GBP"
+    another_placed_order_payment:
+        state: "cart"
```
`placed_order_payment` should be in `new` state

dlobato replied:
I've just updated another_placed_order_payment as placed_order_payment was already there.

lchrusciel commented on src/Handler/Order/UpdatePaymentMethodHandler.php:
```diff
@@ -0,0 +1,50 @@
+
+        $payment = $order->getPayments()[$choosePaymentMethod->paymentIdentifier()];
+
+        $payment->setMethod($paymentMethod);
```
We should also assert that payment is in `new` state

GSadee commented on spec/Validator/Order/OrderExistsValidatorSpec.php:
```diff
@@ -0,0 +1,63 @@
+        ));
+    }
+
+    function it_adds_constraint_if_order_does_not_exits_exists(
```
`it_adds_constraint_if_order_does_not_exist`

lchrusciel commented:
Hello David,

Sorry for such a long response time. I found some time to dig into ShopAPI and recently fix the way, how we handle placed orders in the test environment (https://github.com/Sylius/ShopApiPlugin/pull/439/commits/0e445e4b876e96139e45a53a5e98c8c87fa745f5)

This made this PR conflicted. Would you like to fix it?

dlobato commented:
@lchrusciel I've updated the tests. But now I can't set the payment state to something different from awaiting_payment to test it_does_not_allow_to_update_payment_method_on_paid_order, any suggestion?
I'd say that I need to trigger pay, but not sure how...

mamazu commented:
You can not set the payment method once the payment is completed.

mamazu commented:
Ah, I see your problem. Currently there is no way for the API to mark an order as payed. What you can do is try to generate a fixture with a payed order. Other than that I don't know see #310

jdeveloper commented:
What's the status of this issue?

Is there any hope to get merged? It would be great. So if I choose a payment method, paypal for example, complete the order but then I cancel in paypal, I could change the payment method.

mamazu commented:
I have rebased the issue but I can not push it. @dlobato Please allow edits from maintainers to your branch.

dlobato commented:
@mamazu I don't know how to enable that. I was trying this: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork But I don't see said checkbox anywhere...

mamazu commented:
Hm, very strange. I can push to other merge request. Anyhow I have pushed my version of the branch to my repository [mamazu/handle_order_payment](https://github.com/mamazu/ShopApiPlugin/tree/handle_order_payment). You can run the following commands to get this merge request up to speed:
```bash
# Add me as a remote
git remote add mamazu git@github.com:mamazu/ShopApiPlugin.git

# Fetch all the branches
git fetch mamazu

# If you are not on your local branch change to it. Otherwise just run
git reset --hard mamazu/handle_order_payment

# Push the new changes to git (you might need to explicitly add `origin handle_order_payment`)
git push --force-with-lease
```

dlobato commented:
Your changes are already on my branch...

mamazu commented:
This can't be I have fixed the tests (eg. removed the channelcode from the route).

dlobato commented:
Check it out now

mamazu commented:
Looks good. Can you run `composer fix` and then commit the changes. Then the merge request ready for merge.

mamazu commented on src/Request/Order/UpdatePaymentMethodRequest.php:
```diff
@@ -0,0 +1,42 @@
+        return $this->token;
+    }
+
+    public function getPaymentId()
```
```suggestion
    /** @return int|string */
    public function getPaymentId()
```

mamazu commented on src/Command/Order/UpdatePaymentMethod.php:
```diff
@@ -0,0 +1,39 @@
+    /** @var string */
+    protected $paymentMethodCode;
+
+    public function __construct(string $orderToken, $paymentId, string $paymentMethodCode)
```
```suggestion
    /** @param int|string $paymentId */
    public function __construct(string $orderToken, $paymentId, string $paymentMethodCode)
```

mamazu commented on src/Command/Order/UpdatePaymentMethod.php:
```diff
@@ -0,0 +1,39 @@
+        return $this->orderToken;
+    }
+
+    public function paymentId()
```
```suggestion
    /** @return string|int */
    public function paymentId()
```

mamazu commented on src/Validator/Order/OrderExistsValidator.php:
```diff
@@ -0,0 +1,27 @@
+    public function validate($token, Constraint $constraint): void
+    {
+        if (null === $this->orderRepository->findOneBy(['tokenValue' => $token, 'state' => $constraint->state])) {
+            $this->context->addViolation($constraint->message);
```
```suggestion
            /** @var Sylius\ShopApiPlugin\Validator\Constraints\OrderExists $constraint */
            $this->context->addViolation($constraint->message);
```

mamazu commented on src/Validator/Order/OrderExistsValidator.php:
```diff
@@ -0,0 +1,28 @@
+    public function validate($token, Constraint $constraint): void
+    {
+        if (null === $this->orderRepository->findOneBy(['tokenValue' => $token, 'state' => $constraint->state])) {
+            /** @var Sylius\ShopApiPlugin\Validator\Constraints\OrderExists $constraint */
```
```suggestion
```

mamazu commented on src/Validator/Order/OrderExistsValidator.php:
```diff
@@ -0,0 +1,28 @@
+
+    public function validate($token, Constraint $constraint): void
+    {
+        if (null === $this->orderRepository->findOneBy(['tokenValue' => $token, 'state' => $constraint->state])) {
```
```suggestion
        /** @var Sylius\ShopApiPlugin\Validator\Constraints\OrderExists $constraint */
        if (null === $this->orderRepository->findOneBy(['tokenValue' => $token, 'state' => $constraint->state])) {
```

mamazu commented:
Can you please checkout the code and put the right doc comment in there.

mamazu commented:
Thank you, David! :1st_place_medal:

lchrusciel commented on src/Handler/Order/UpdatePaymentMethodHandler.php:
```diff
@@ -0,0 +1,52 @@
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $choosePaymentMethod->orderToken()]);
+
+        Assert::notNull($order, 'Order has not been found.');
+        Assert::notSame(OrderInterface::STATE_CART, $order->getState(), 'Only orders can be updated.');
```
Not 100% true, I suppose. Order's in fulfilled state should be blocked as well. 

```suggestion
        Assert::same(OrderInterface::STATE_NEW, $order->getState(), 'Only new orders can be updated.');
```

mamazu replied:
The validator checks if the payment you are trying to edit was not payed yet. This would be a safer option as an assertion?
